src/core/lib/resource_quota/memory_quota.cc

```diff
@@ -0,0 +1,378 @@
+      Ref(DEBUG_LOCATION, "post_reclaimer"), std::move(fn));
+}
+
+//
+// AtomicBarrier
+//
+
+Poll<AtomicBarrier::WaitPromise::Empty>
+AtomicBarrier::WaitPromise::operator()() {
+  if (barrier_->counter_.load(std::memory_order_acquire) != token_) {
+    return Empty{};
+  }
+  barrier_->waker_ = Activity::current()->MakeNonOwningWaker();
+  return Pending{};
+}
+
+uint64_t AtomicBarrier::NewToken() {
+  return counter_.fetch_add(1, std::memory_order_relaxed);
+}
+
+void AtomicBarrier::Notify(uint64_t token) {
+  if (counter_.compare_exchange_strong(token, token + 1,
+                                       std::memory_order_acq_rel,
+                                       std::memory_order_relaxed)) {
+    waker_.Wakeup();
+  }
+}
+
+//
+// MemoryQuota
+//
+
+MemoryQuota::MemoryQuota() {
+  auto self = WeakRef();
+
+  auto reclamation_loop = Loop(Seq(
+      [self]() -> Poll<int> {
+        // If there's free memory we no longer need to reclaim memory!
+        if (self->free_bytes_.load(std::memory_order_acquire) > 0) {
+          return Pending{};
+        }
+        return 0;
+      },
+      [self]() {
+        // Race biases to the first thing that completes... so this will
+        // choose the highest priority/least destructive thing to do that's
+        // available.
+        return Race(self->reclaimers_[0].Next(), self->reclaimers_[1].Next(),
+                    self->reclaimers_[2].Next(), self->reclaimers_[3].Next());
+      },
+      [self](ReclamationFunction reclaimer) {
+        // One of the reclaimer queues gave us a way to get back memory.
+        // Call the reclaimer with a token that contains enough to wake us
+        // up again.
+        uint64_t token = self->barrier_.NewToken();
+        reclaimer(ReclamationSweep(self, token));
+        // Return a promise that will wait for our barrier. This will be
+        // awoken by the token above being destroyed. So, once that token is
+        // destroyed, we'll be able to proceed.
+        return self->barrier_.Wait(token);
+      },
+      []() -> LoopCtl<absl::Status> {
+        // Continue the loop!
+        return Continue{};
+      }));
+
+  reclaimer_activity_ = MakeActivity(
+      std::move(reclamation_loop),
+      [](std::function<void()> f) { std::thread(f).detach(); },
+      [](absl::Status status) {
+        GPR_ASSERT(status.code() == absl::StatusCode::kCancelled);
+      });
+}
+
+void MemoryQuota::SetSize(size_t new_size) {
+  size_t old_size = quota_size_.exchange(new_size, std::memory_order_relaxed);
+  if (old_size < new_size) {
+    // We're growing the quota.
+    Return(new_size - old_size);
+  } else {
+    // We're shrinking the quota.
+    Take(old_size - new_size);
+  }
+}
+
+void MemoryQuota::Take(size_t amount) {
+  // If there's a request for nothing, then do nothing!
+  if (amount == 0) return;
+  // Grab memory from the quota.
+  auto prior = free_bytes_.fetch_sub(amount, std::memory_order_acq_rel);
```

Comment:
Is it worth adding debug-assert that this doesn't underflow?

Reply:
This one is supposed to underflow... probably wants a comment somewhere.

Basically, we succeed every allocation, and run a loop in the background:
```
while (free_bytes_ <= 0) reclaim_some_stuff();
```

Reply:
(I expect to add some broad 'don't accept new {calls,connections}' logic soon too when memory pressure is high, but acceptance of high level things needs a different api than allocation.